keps/sig-node/4381-dra-numeric-parameters/README.md

```diff
@@ -0,0 +1,1054 @@
+behalf of the driver. This implies that kubelet, DRA driver, and apiserver must
+all support the same semantic models and the same versions of them. It might be
+possible to support version skew (= keeping kubelet at an older version than
+the control plane and the DRA drivers) in the future, but currently this is out
+of scope.
+
+### Using semantic parameters as claim parameters
+
+The semantic model also defines a type for claim parameters. For the example
+above, that could be:
+```
+type Parameters struct {
+	Required int64
+}
+```
+
+The `Parameters.Required` value gets compared against `Attributes.Count`.
+
+Suppose a vendor has a CRD which allows requesting a certain number of `frobnicators`:
+
+```
+kind: ClaimParameters
+apiVersion: dra.frobincatorInc.example.com/v1alpha1
+metadata:
+  name: my-claim-parameters
+  uid: abcd
+  resourceVersion: 123
+spec:
+  requiredFrobincators: 100
+```
+
+The vendor CRD controller needs to sync this CRs into one in-tree object:
+```
+kind: ResourceClaimParameters
+apiVersion: resource.k8s.io/v1alpha2
+metadata:
+  name: frobnicator-claim-parameters
+  namespace: default
+generatedFrom:
+  kind: ClaimParameters
+  apiGroup: dra.frobincatorInc.example.com
+  name: my-claim-parameters
+  resourceVersion: 123
+  uid: abcd
+parameters:
+  - counter:
+      required: 100
+```
+
+### Notes/Constraints/Caveats
+
+For this proposal to work, kube-scheduler must have permission to read, list
+and watch the custom claim parameter resources. This is necessary because it is
+not known in advance which CRDs will be used by DRA drivers. When deploying a
+driver, suitable role bindings for
+[`system:kube-scheduler`](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#core-component-roles)
+have to be added which grant read, watch and list permissions for the vendor
+CRD.
+
+The scheduler does not need to interpret the custom data in those CRDs. It
+still needs to read them for two purposes:
+- Ensure that the in-tree claim parameters match the vendor claim parameters
+  reference by the claim and are not out-dated.
+- Include a snapshot of claim and class parameters in the allocation result,
+  to ensure that configuration parameters are available when the DRA driver
+  needs them.
+
+The types that the scheduler needs to watch may change at runtime as classes
+get added or removed. To avoid repeated GET operations for those CRDs, dynamic
+informers will have to be used that get started and stopped as needed.
+
+## Design Details
+
+### ResourceClass extension
+
+A new, optional field in ResourceClass enables semantic parameters for claims
+using this class:
+
+```
+type ResourceClass struct {
+	...
+
+	// If (and only if) allocation of claims using this class is handled
+	// via numeric parameters, then ParameterTypes describes those
+	// parameters. A driver might support multiple different CRDs for
+	// its claims.
+	ParameterTypes []SemanticParameterType
```

Comment:
I don't understand why we need this. Now that we don't do it through CEL or templates, I think this information would be embedded in the code of the DRA driver controller that translates the CR into the common parameter model.

Reply:
As discussed in https://github.com/kubernetes/enhancements/pull/4384#discussion_r1477019579 and https://github.com/kubernetes/enhancements/pull/4384#discussion_r1476910024 we still need to know about vendor CRDs.